src/public/modules/forms/helpers/response-factory.ts

```diff
@@ -0,0 +1,65 @@
+import { hasProp } from '../../../../shared/util/has-prop'
+import {
+  ArrayResponse,
+  DisplayedResponseWithoutAnswer,
+  NestedResponse,
+  SingleResponse,
+} from '../../../../types/response'
+
+import {
+  ArrayAnswerResponse,
+  Response,
+  SingleAnswerResponse,
+  TableResponse,
+} from './csv-response-classes'
+
+export const getResponseInstance = (
+  fieldRecordData: DisplayedResponseWithoutAnswer,
+): Response => {
+  console.log(fieldRecordData)
+  if (
+    isNestedResponse(fieldRecordData) &&
+    fieldRecordData.fieldType === 'table'
+  ) {
+    return new TableResponse(fieldRecordData)
+  } else if (
+    isArrayResponse(fieldRecordData) &&
+    fieldRecordData.fieldType === 'checkbox'
+  ) {
+    return new ArrayAnswerResponse(fieldRecordData)
+  } else if (isSingleResponse(fieldRecordData)) {
+    return new SingleAnswerResponse(fieldRecordData)
+  } else {
+    // eslint-disable-next-line typesafe/no-throw-sync-func
+    throw new Error('Response did not match any known type') // should be caught in submissions client factory
+  }
+}
+
+const isNestedResponse = (
+  response: DisplayedResponseWithoutAnswer,
+): response is NestedResponse => {
+  return (
+    hasProp(response, 'answerArray') &&
```

Comment:
wdyt about extracting 

```ts
hasProp(response, 'answerArray') &&
Array.isArray(response.answerArray)
```

into its own function (maybe `isAnswerArray`) and then having the additional subtyping checks?

Reply:
Hm.. I know I won't be able to do this with a normal `isAnswerArray` function that returns a boolean as typescript will complain that `answerArray` may not be defined. Do you know of something similar to a type guard that can help to overcome this issue?

Reply:
i think you'd need to create a base type! so the base function is a type predicate for that then the other nested ones just extend the base type so ts can infer properly!